### ISSUE TITLE
Accepting and getting connections on the threading server without launching a new thread

### ISSUE THREAD

**Issue by Shir0kamii** [python-websockets/websockets#1421]
Hi,

In the company I work for, we are building a software that acts as a man-in-the-middle between some hardware and a server it connects to. We plan to write integrations tests using your library. The tests we would like to write with pytest would look like this:

```python
def test_anything(server_connection, client_connection):
    client_connection.send("foo")
    msg = server_connection.recv()
    # test how our software affected the message between client and server
    assert msg == "foo123"
```

For this, we need to be able to accept a connection on the websocket server without spinning a new thread with a handler.

We are planning to fork your library and implement that feature on our side. I could open a MR if this is something that you are interested in integrating, but if so, I would like to know how you would see it done.

API-wise, i'm thinking something like this would look fine:

```python
# Notice we don't pass a handler since we won't need it
# We could also pass a dumb handler if we want to keep it mandatory
with websocket.sync.server.serve(host="localhost", port=8000) as server:
    server_connection = server.accept()
```

For this to work, the websocket subprotocol selection and handshake would need to be done outside the request handler, maybe in the new `accept` method, which would also contain the polling code. The `serve_forever` method would then be calling `accept` and launching threads with the provided custom handler and the accepted websocket connection.

What do you think about that? Do you think other people could have the same use case?

**TL;DR**: We will fork your library to add the feature to accept connections without launching a handler in a new thread. We are willing to upstream the feature, so discussing how to implement it ahead of time would be better if it interest you.

PS: This is for the treading server, but the same could be done for the asyncio server in a separate PR on my own time.

**Comment by aaugustin** [comment]
Or, if you want to go low level, you could also use the Sans-I/O layer for testing.

Docs here: https://websockets.readthedocs.io/en/stable/howto/sansio.html

Essentially you'd receive a TCP connection from your hardware and pass them to the Sans-I/O layer. Perhaps you can have a minimalistic bridge that's good enough for your tests. That would likely still be less work than forking this library.

**Comment by aaugustin** [comment]
Wait - I have a far simpler idea - instantiate directly a [ServerConnection](https://websockets.readthedocs.io/en/stable/reference/sync/server.html#websockets.sync.server.ServerConnection) - that should just work?

```python
import socket
import websockets.server
import websockets.sync.server

sock = socket.create_server((HOST, PORT))
conn, addr = sock.accept()
server_connection = websockets.sync.server.ServerConnection(conn, websockets.server.ServerProtocol())
```

At that point, you can do what you want, I think.

You just bypassed all the server machinery by accepting a server connection.

Also I think it's going to be much easier to make that version robust; cleanup is as simple as:

```python
server_connection.close()
conn.close()
sock.close()
```

(+ exception handling)

**Comment by aaugustin** [comment]
If that last option works, I'll add it to the FAQ.

**Comment by Shir0kamii** [comment]
Indeed, the last option is what would work best in our case, thanks for the suggestion!

**Comment by aaugustin** [comment]
As they say - third time is the charm :-)

**Comment by aaugustin** [comment]
Upon further thought, this is too much of an edge case to document in the FAQ.
